File: 23.NoOfWordsinTrie/NoOfWordsinTrie.java
import java.util.*;

class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isEnd = false;
}

class Trie {
    TrieNode root = new TrieNode();
    int count = 0;

    void insert(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            node.children.putIfAbsent(c, new TrieNode());
            node = node.children.get(c);
        }
        if (!node.isEnd) {
            node.isEnd = true;
            count++;
        }
    }

    int totalWords() {
        return count;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = Integer.parseInt(sc.nextLine());
        Trie trie = new Trie();
        for (int i = 0; i < n; i++) {
            String word = sc.nextLine();
            trie.insert(word);
        }
        System.out.println(trie.totalWords());
    }
}

File: 23.NoOfWordsinTrie/NoOfWordsinTrie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
        self.count = 0

    def insert(self, word):
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        if not node.isEnd:
            node.isEnd = True
            self.count += 1

    def totalWords(self):
        return self.count

# Main
n = int(input())
trie = Trie()
for _ in range(n):
    word = input().strip()
    trie.insert(word)
print(trie.totalWords())
